# Welcome!

# 1. 기초 라이브러리 명령어 학습

## 1.2 pandas 라이브러리

Pandas는 열과 행으로 이루어져 있는 Table을 처리하고 가공하는 라이브러리로, Python으로 데이터를 다룰 때 빠질 수 없는 아주 중요한 패키지입니다. 한 열로 구성된 1차원 형태의 Series, 열들이 모여 Table을 이루는 DataFrame, 2가지 자료형으로 구성되어 있습니다.

<img src="https://dandyrilla.github.io/images/2017-08-12/fig0.png" width="400" height="400" />

In [7]:
import numpy as np
import pandas as pd

### Series와 DataFrame 기초

#### Series

In [11]:
# Series는 1차원 배열로 실제 값들이 나열되어 있는 Value와 이러한 값들에 번호를 매기는 Index로 구성되어 있습니다.
# index가 따로 지정되지 않으면, 자동으로 0부터 순차적으로 지정됩니다
a = [1,2,3,4,5]
a_series = pd.Series(a)
print(a)
print(a_series)

[1, 2, 3, 4, 5]
0    1
1    2
2    3
3    4
4    5
dtype: int64


In [13]:
# 인덱스 지정
b = [6,7,8,9,10]
b_index = ['a', 'b', 'c', 'd', 'e']
b_series = pd.Series(b, index=b_index)
print(b)
print(b_series)

[6, 7, 8, 9, 10]
a     6
b     7
c     8
d     9
e    10
dtype: int64


In [15]:
# array와 마찬가지로 대괄호 안에 Index 값을 넣어 특정 값만 추출할 수 있습니다
print(a_series[2])
print(b_series['b'])

3
7


## DataFrame

In [18]:
# DataFrame은 2치원 형태의 자료구조이며, 행은 Index 번호로, 열은 column 이름으로 구분합니다.
raw_data = {'col0': [1,2,3,4,5],
           'col1': [10,20,30,40,50],
           'col2': [100,200,300,400,500]}
df = pd.DataFrame(raw_data)
df

col0  col1  col2
0     1    10   100
1     2    20   200
2     3    30   300
3     4    40   400
4     5    50   500

In [21]:
# DataFrame의 각 열은 Series가 됩니다.
# DataFrame에서 Series만 추가하는 방식은 대괄호 방식, 속성 추출방식이 있습니다.
print(df['col0'])
print(df.col1)
print(type(df))
print(type(df.col1))

0    1
1    2
2    3
3    4
4    5
Name: col0, dtype: int64
0    10
1    20
2    30
3    40
4    50
Name: col1, dtype: int64
<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.series.Series'>


In [23]:
# Dataframe에 열을 추가할 수 있습니다.
new=['딸기', '당근', '수박', '참외', '메론']
df['new column'] = new
df

col0  col1  col2 new column
0     1    10   100         딸기
1     2    20   200         당근
2     3    30   300         수박
3     4    40   400         참외
4     5    50   500         메론

In [25]:
# del 명렁어로 열을 삭제할 수도 있습니다
del df['col2']
df

col0  col1 new column
0     1    10         딸기
1     2    20         당근
2     3    30         수박
3     4    40         참외
4     5    50         메론

In [27]:
# drop 명렁어로 열을 제외할때는 axis=1 명시
df.drop('new column', axis=1)

col0  col1
0     1    10
1     2    20
2     3    30
3     4    40
4     5    50

### loc & iloc

In [30]:
# loc 명령어를 통해 DataFrame의 특정 행, 특정 열을 추출할 수 있습니다
df.loc[3]     #특정 행만 추출

col0           4
col1          40
new column    참외
Name: 3, dtype: object

In [32]:
df.loc[1,['col0']]

col0    2
Name: 1, dtype: object

In [34]:
df.loc[1:4,:] #특정 행과 모든 열

col0  col1 new column
1     2    20         당근
2     3    30         수박
3     4    40         참외
4     5    50         메론

In [36]:
df.loc[:,['col0','new column']] #특정 열과 모든 행

col0 new column
0     1         딸기
1     2         당근
2     3         수박
3     4         참외
4     5         메론

In [38]:
df.loc[3:5,['col1','new column']]  #특정 열과 특정 행

col1 new column
3    40         참외
4    50         메론

In [40]:
df.loc[[1,2,4],['col1','new column']] #특정 열과 특정 행

col1 new column
1    20         당근
2    30         수박
4    50         메론

In [42]:
# 컬럼 순서 변경
df.loc[:,['new column', 'col1', 'col0']]

new column  col1  col0
0         딸기    10     1
1         당근    20     2
2         수박    30     3
3         참외    40     4
4         메론    50     5

In [44]:
df.index = ['a','b','c','d','e']    #index를 숫자에서 문자로 바꾸었습니다.
df.loc[['b','d'],['new column']]  # loc은 index와 column 값 기준

new column
b         당근
d         참외

In [46]:
#iloc은 실제 배열 순서 기준으로 통해 DataFrame의 특정 행, 특정 열을 추출할 수 있습니다

In [48]:
df.iloc[[1,3],[2]]      #iloc은 실제 배열 순서 기준

new column
b         당근
d         참외

In [50]:
# DataFrame도 전치(행과 열을 바꿈)할 수 있습니다
df.T

a   b   c   d   e
col0         1   2   3   4   5
col1        10  20  30  40  50
new column  딸기  당근  수박  참외  메론

In [52]:
df

col0  col1 new column
a     1    10         딸기
b     2    20         당근
c     3    30         수박
d     4    40         참외
e     5    50         메론

In [54]:
(df['col1'] <= 30) & (df['new column'] == '딸기')

a     True
b    False
c    False
d    False
e    False
dtype: bool

In [56]:
# col1 30도 이하이고, new column 딸기이면, 산다 논다는 데이터를 입력해봅시다.
df['buy'] = (df['col1'] <= 30) & (df['new column'] == '딸기')
df

col0  col1 new column    buy
a     1    10         딸기   True
b     2    20         당근  False
c     3    30         수박  False
d     4    40         참외  False
e     5    50         메론  False

## Import data

In [60]:
# pd.set_option('display.max_columns', None)

# pd.set_option('display.max_rows', None)
# pd.set_option('display.max_rows', 60)

In [64]:
# Colab 사용
#flight_data = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/5_스마트도시데이터분석/codes/lecture02_data_processing/data/nycflights13.csv')

# Jupyter Notebook 사용
flight_data = pd.read_csv('data/nycflights13.csv')


flight_data

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     533.0             529        4.0     850.0   
2       2013      1    1     542.0             540        2.0     923.0   
3       2013      1    1     544.0             545       -1.0    1004.0   
4       2013      1    1     554.0             600       -6.0     812.0   
...      ...    ...  ...       ...             ...        ...       ...   
336771  2013      9   30       NaN            1455        NaN       NaN   
336772  2013      9   30       NaN            2200        NaN       NaN   
336773  2013      9   30       NaN            1210        NaN       NaN   
336774  2013      9   30       NaN            1159        NaN       NaN   
336775  2013      9   30       NaN             840        NaN       NaN   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
0                  819       11.0      UA    1545  N14228    EWR  IAH   
1                  830       20.0      UA    1714  N24211    LGA  IAH   
2                  850       33.0      AA    1141  N619AA    JFK  MIA   
3                 1022      -18.0      B6     725  N804JB    JFK  BQN   
4                  837      -25.0      DL     461  N668DN    LGA  ATL   
...                ...        ...     ...     ...     ...    ...  ...   
336771            1634        NaN      9E    3393     NaN    JFK  DCA   
336772            2312        NaN      9E    3525     NaN    LGA  SYR   
336773            1330        NaN      MQ    3461  N535MQ    LGA  BNA   
336774            1344        NaN      MQ    3572  N511MQ    LGA  CLE   
336775            1020        NaN      MQ    3531  N839MQ    LGA  RDU   

        air_time  distance  hour  minute            time_hour  
0          227.0      1400     5      15  2013-01-01 05:00:00  
1          227.0      1416     5      29  2013-01-01 05:00:00  
2          160.0      1089     5      40  2013-01-01 05:00:00  
3          183.0      1576     5      45  2013-01-01 05:00:00  
4          116.0       762     6       0  2013-01-01 06:00:00  
...          ...       ...   ...     ...                  ...  
336771       NaN       213    14      55  2013-09-30 14:00:00  
336772       NaN       198    22       0  2013-09-30 22:00:00  
336773       NaN       764    12      10  2013-09-30 12:00:00  
336774       NaN       419    11      59  2013-09-30 11:00:00  
336775       NaN       431     8      40  2013-09-30 08:00:00  

[336776 rows x 19 columns]

## Select columns
Select the `origin`, `dest`, and `hour` columns

In [67]:
flight_data.loc[:, ['origin', 'dest', 'hour']]

origin dest  hour
0         EWR  IAH     5
1         LGA  IAH     5
2         JFK  MIA     5
3         JFK  BQN     5
4         LGA  ATL     6
...       ...  ...   ...
336771    JFK  DCA    14
336772    LGA  SYR    22
336773    LGA  BNA    12
336774    LGA  CLE    11
336775    LGA  RDU     8

[336776 rows x 3 columns]

In [69]:
# Pandas
flight_data.loc[0:4, ['origin', 'dest', 'hour']]

origin dest  hour
0    EWR  IAH     5
1    LGA  IAH     5
2    JFK  MIA     5
3    JFK  BQN     5
4    LGA  ATL     6

In [71]:
# 새로운 데이터프레임 생성
new_df = flight_data.loc[0:4, ['origin', 'dest', 'hour']]

new_df

origin dest  hour
0    EWR  IAH     5
1    LGA  IAH     5
2    JFK  MIA     5
3    JFK  BQN     5
4    LGA  ATL     6

## Drop columns

In [74]:
flight_subset = flight_data.drop(['origin', 'dest', 'hour'], axis=1).head()
flight_subset

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     542.0             540        2.0     923.0   
3  2013      1    1     544.0             545       -1.0    1004.0   
4  2013      1    1     554.0             600       -6.0     812.0   

   sched_arr_time  arr_delay carrier  flight tailnum  air_time  distance  \
0             819       11.0      UA    1545  N14228     227.0      1400   
1             830       20.0      UA    1714  N24211     227.0      1416   
2             850       33.0      AA    1141  N619AA     160.0      1089   
3            1022      -18.0      B6     725  N804JB     183.0      1576   
4             837      -25.0      DL     461  N668DN     116.0       762   

   minute            time_hour  
0      15  2013-01-01 05:00:00  
1      29  2013-01-01 05:00:00  
2      40  2013-01-01 05:00:00  
3      45  2013-01-01 05:00:00  
4       0  2013-01-01 06:00:00

## Filter rows
Select all flights longer than 10 hours that flew from JFK airport on January 1

In [77]:
# AND
flight_data[(flight_data['month'] == 1) &
            (flight_data['day'] == 1) &
            (flight_data['origin'] == 'JFK') &
            (flight_data['hour'] > 10)]

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
151  2013      1    1     848.0            1835      853.0    1001.0   
258  2013      1    1    1059.0            1100       -1.0    1210.0   
265  2013      1    1    1111.0            1115       -4.0    1222.0   
266  2013      1    1    1112.0            1100       12.0    1440.0   
272  2013      1    1    1124.0            1100       24.0    1435.0   
..    ...    ...  ...       ...             ...        ...       ...   
832  2013      1    1    2326.0            2130      116.0     131.0   
833  2013      1    1    2327.0            2250       37.0      32.0   
835  2013      1    1    2353.0            2359       -6.0     425.0   
836  2013      1    1    2353.0            2359       -6.0     418.0   
837  2013      1    1    2356.0            2359       -3.0     425.0   

     sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
151            1950      851.0      MQ    3944  N942MQ    JFK  BWI      41.0   
258            1215       -5.0      MQ    3792  N509MQ    JFK  DCA      50.0   
265            1226       -4.0      B6      24  N279JB    JFK  BTV      52.0   
266            1438        2.0      UA     285  N517UA    JFK  SFO     364.0   
272            1431        4.0      B6     641  N590JB    JFK  SFO     349.0   
..              ...        ...     ...     ...     ...    ...  ...       ...   
832              18       73.0      B6     199  N594JB    JFK  LAS     290.0   
833            2359       33.0      B6      22  N639JB    JFK  SYR      45.0   
835             445      -20.0      B6     739  N591JB    JFK  PSE     195.0   
836             442      -24.0      B6     707  N794JB    JFK  SJU     185.0   
837             437      -12.0      B6     727  N588JB    JFK  BQN     186.0   

     distance  hour  minute            time_hour  
151       184    18      35  2013-01-01 18:00:00  
258       213    11       0  2013-01-01 11:00:00  
265       266    11      15  2013-01-01 11:00:00  
266      2586    11       0  2013-01-01 11:00:00  
272      2586    11       0  2013-01-01 11:00:00  
..        ...   ...     ...                  ...  
832      2248    21      30  2013-01-01 21:00:00  
833       209    22      50  2013-01-01 22:00:00  
835      1617    23      59  2013-01-01 23:00:00  
836      1598    23      59  2013-01-01 23:00:00  
837      1576    23      59  2013-01-01 23:00:00  

[213 rows x 19 columns]

In [79]:
# OR
flight_data[(flight_data['month'] == 1) |
            (flight_data['origin'] == 'JFK') |
            (flight_data['origin'] == 'LGA')]

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     533.0             529        4.0     850.0   
2       2013      1    1     542.0             540        2.0     923.0   
3       2013      1    1     544.0             545       -1.0    1004.0   
4       2013      1    1     554.0             600       -6.0     812.0   
...      ...    ...  ...       ...             ...        ...       ...   
336771  2013      9   30       NaN            1455        NaN       NaN   
336772  2013      9   30       NaN            2200        NaN       NaN   
336773  2013      9   30       NaN            1210        NaN       NaN   
336774  2013      9   30       NaN            1159        NaN       NaN   
336775  2013      9   30       NaN             840        NaN       NaN   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
0                  819       11.0      UA    1545  N14228    EWR  IAH   
1                  830       20.0      UA    1714  N24211    LGA  IAH   
2                  850       33.0      AA    1141  N619AA    JFK  MIA   
3                 1022      -18.0      B6     725  N804JB    JFK  BQN   
4                  837      -25.0      DL     461  N668DN    LGA  ATL   
...                ...        ...     ...     ...     ...    ...  ...   
336771            1634        NaN      9E    3393     NaN    JFK  DCA   
336772            2312        NaN      9E    3525     NaN    LGA  SYR   
336773            1330        NaN      MQ    3461  N535MQ    LGA  BNA   
336774            1344        NaN      MQ    3572  N511MQ    LGA  CLE   
336775            1020        NaN      MQ    3531  N839MQ    LGA  RDU   

        air_time  distance  hour  minute            time_hour  
0          227.0      1400     5      15  2013-01-01 05:00:00  
1          227.0      1416     5      29  2013-01-01 05:00:00  
2          160.0      1089     5      40  2013-01-01 05:00:00  
3          183.0      1576     5      45  2013-01-01 05:00:00  
4          116.0       762     6       0  2013-01-01 06:00:00  
...          ...       ...   ...     ...                  ...  
336771       NaN       213    14      55  2013-09-30 14:00:00  
336772       NaN       198    22       0  2013-09-30 22:00:00  
336773       NaN       764    12      10  2013-09-30 12:00:00  
336774       NaN       419    11      59  2013-09-30 11:00:00  
336775       NaN       431     8      40  2013-09-30 08:00:00  

[225834 rows x 19 columns]

## Working with Nan

In [82]:
flight_data_nan = flight_data
flight_data_nan.dtypes

year                int64
month               int64
day                 int64
dep_time          float64
sched_dep_time      int64
dep_delay         float64
arr_time          float64
sched_arr_time      int64
arr_delay         float64
carrier            object
flight              int64
tailnum            object
origin             object
dest               object
air_time          float64
distance            int64
hour                int64
minute              int64
time_hour          object
dtype: object

In [84]:
# Check observation count
flight_data_nan.count()

year              336776
month             336776
day               336776
dep_time          328521
sched_dep_time    336776
dep_delay         328521
arr_time          328063
sched_arr_time    336776
arr_delay         327346
carrier           336776
flight            336776
tailnum           334264
origin            336776
dest              336776
air_time          327346
distance          336776
hour              336776
minute            336776
time_hour         336776
dtype: int64

In [86]:
flight_data_nan['arr_delay'].isna()

0         False
1         False
2         False
3         False
4         False
          ...  
336771     True
336772     True
336773     True
336774     True
336775     True
Name: arr_delay, Length: 336776, dtype: bool

In [88]:
# 결측치 있는 행만 골라내기
flight_data_nan.loc[flight_data_nan['arr_delay'].isna(), :]

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
471     2013      1    1    1525.0            1530       -5.0    1934.0   
477     2013      1    1    1528.0            1459       29.0    2002.0   
615     2013      1    1    1740.0            1745       -5.0    2158.0   
643     2013      1    1    1807.0            1738       29.0    2251.0   
725     2013      1    1    1939.0            1840       59.0      29.0   
...      ...    ...  ...       ...             ...        ...       ...   
336771  2013      9   30       NaN            1455        NaN       NaN   
336772  2013      9   30       NaN            2200        NaN       NaN   
336773  2013      9   30       NaN            1210        NaN       NaN   
336774  2013      9   30       NaN            1159        NaN       NaN   
336775  2013      9   30       NaN             840        NaN       NaN   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
471               1805        NaN      MQ    4525  N719MQ    LGA  XNA   
477               1647        NaN      EV    3806  N17108    EWR  STL   
615               2020        NaN      MQ    4413  N739MQ    LGA  XNA   
643               2103        NaN      UA    1228  N31412    EWR  SAN   
725               2151        NaN      9E    3325  N905XJ    JFK  DFW   
...                ...        ...     ...     ...     ...    ...  ...   
336771            1634        NaN      9E    3393     NaN    JFK  DCA   
336772            2312        NaN      9E    3525     NaN    LGA  SYR   
336773            1330        NaN      MQ    3461  N535MQ    LGA  BNA   
336774            1344        NaN      MQ    3572  N511MQ    LGA  CLE   
336775            1020        NaN      MQ    3531  N839MQ    LGA  RDU   

        air_time  distance  hour  minute            time_hour  
471          NaN      1147    15      30  2013-01-01 15:00:00  
477          NaN       872    14      59  2013-01-01 14:00:00  
615          NaN      1147    17      45  2013-01-01 17:00:00  
643          NaN      2425    17      38  2013-01-01 17:00:00  
725          NaN      1391    18      40  2013-01-01 18:00:00  
...          ...       ...   ...     ...                  ...  
336771       NaN       213    14      55  2013-09-30 14:00:00  
336772       NaN       198    22       0  2013-09-30 22:00:00  
336773       NaN       764    12      10  2013-09-30 12:00:00  
336774       NaN       419    11      59  2013-09-30 11:00:00  
336775       NaN       431     8      40  2013-09-30 08:00:00  

[9430 rows x 19 columns]

In [90]:
# 결측치 없는 행만 골라내기
flight_data_nan.loc[flight_data_nan['arr_delay'].notna(), :]

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     533.0             529        4.0     850.0   
2       2013      1    1     542.0             540        2.0     923.0   
3       2013      1    1     544.0             545       -1.0    1004.0   
4       2013      1    1     554.0             600       -6.0     812.0   
...      ...    ...  ...       ...             ...        ...       ...   
336765  2013      9   30    2240.0            2245       -5.0    2334.0   
336766  2013      9   30    2240.0            2250      -10.0    2347.0   
336767  2013      9   30    2241.0            2246       -5.0    2345.0   
336768  2013      9   30    2307.0            2255       12.0    2359.0   
336769  2013      9   30    2349.0            2359      -10.0     325.0   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
0                  819       11.0      UA    1545  N14228    EWR  IAH   
1                  830       20.0      UA    1714  N24211    LGA  IAH   
2                  850       33.0      AA    1141  N619AA    JFK  MIA   
3                 1022      -18.0      B6     725  N804JB    JFK  BQN   
4                  837      -25.0      DL     461  N668DN    LGA  ATL   
...                ...        ...     ...     ...     ...    ...  ...   
336765            2351      -17.0      B6    1816  N354JB    JFK  SYR   
336766               7      -20.0      B6    2002  N281JB    JFK  BUF   
336767               1      -16.0      B6     486  N346JB    JFK  ROC   
336768            2358        1.0      B6     718  N565JB    JFK  BOS   
336769             350      -25.0      B6     745  N516JB    JFK  PSE   

        air_time  distance  hour  minute            time_hour  
0          227.0      1400     5      15  2013-01-01 05:00:00  
1          227.0      1416     5      29  2013-01-01 05:00:00  
2          160.0      1089     5      40  2013-01-01 05:00:00  
3          183.0      1576     5      45  2013-01-01 05:00:00  
4          116.0       762     6       0  2013-01-01 06:00:00  
...          ...       ...   ...     ...                  ...  
336765      41.0       209    22      45  2013-09-30 22:00:00  
336766      52.0       301    22      50  2013-09-30 22:00:00  
336767      47.0       264    22      46  2013-09-30 22:00:00  
336768      33.0       187    22      55  2013-09-30 22:00:00  
336769     196.0      1617    23      59  2013-09-30 23:00:00  

[327346 rows x 19 columns]

In [92]:
# Filling missing
flight_data_nan['arr_delay'] = flight_data_nan['arr_delay'].fillna(0)

In [94]:
flight_data_nan.count()

year              336776
month             336776
day               336776
dep_time          328521
sched_dep_time    336776
dep_delay         328521
arr_time          328063
sched_arr_time    336776
arr_delay         336776
carrier           336776
flight            336776
tailnum           334264
origin            336776
dest              336776
air_time          327346
distance          336776
hour              336776
minute            336776
time_hour         336776
dtype: int64

In [96]:
flight_data

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     533.0             529        4.0     850.0   
2       2013      1    1     542.0             540        2.0     923.0   
3       2013      1    1     544.0             545       -1.0    1004.0   
4       2013      1    1     554.0             600       -6.0     812.0   
...      ...    ...  ...       ...             ...        ...       ...   
336771  2013      9   30       NaN            1455        NaN       NaN   
336772  2013      9   30       NaN            2200        NaN       NaN   
336773  2013      9   30       NaN            1210        NaN       NaN   
336774  2013      9   30       NaN            1159        NaN       NaN   
336775  2013      9   30       NaN             840        NaN       NaN   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
0                  819       11.0      UA    1545  N14228    EWR  IAH   
1                  830       20.0      UA    1714  N24211    LGA  IAH   
2                  850       33.0      AA    1141  N619AA    JFK  MIA   
3                 1022      -18.0      B6     725  N804JB    JFK  BQN   
4                  837      -25.0      DL     461  N668DN    LGA  ATL   
...                ...        ...     ...     ...     ...    ...  ...   
336771            1634        0.0      9E    3393     NaN    JFK  DCA   
336772            2312        0.0      9E    3525     NaN    LGA  SYR   
336773            1330        0.0      MQ    3461  N535MQ    LGA  BNA   
336774            1344        0.0      MQ    3572  N511MQ    LGA  CLE   
336775            1020        0.0      MQ    3531  N839MQ    LGA  RDU   

        air_time  distance  hour  minute            time_hour  
0          227.0      1400     5      15  2013-01-01 05:00:00  
1          227.0      1416     5      29  2013-01-01 05:00:00  
2          160.0      1089     5      40  2013-01-01 05:00:00  
3          183.0      1576     5      45  2013-01-01 05:00:00  
4          116.0       762     6       0  2013-01-01 06:00:00  
...          ...       ...   ...     ...                  ...  
336771       NaN       213    14      55  2013-09-30 14:00:00  
336772       NaN       198    22       0  2013-09-30 22:00:00  
336773       NaN       764    12      10  2013-09-30 12:00:00  
336774       NaN       419    11      59  2013-09-30 11:00:00  
336775       NaN       431     8      40  2013-09-30 08:00:00  

[336776 rows x 19 columns]

## Sort rows

In [99]:
# Pandas
flight_data.sort_values(by=['distance', 'hour'], ascending=True).head()

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
275945  2013      7   27       NaN             106        NaN       NaN   
3083    2013      1    4    1240.0            1200       40.0    1333.0   
3901    2013      1    5    1155.0            1200       -5.0    1241.0   
3426    2013      1    4    1829.0            1615      134.0    1937.0   
10235   2013      1   12    1613.0            1617       -4.0    1708.0   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
275945             245        0.0      US    1632     NaN    EWR  LGA   
3083              1306       27.0      EV    4193  N14972    EWR  PHL   
3901              1306      -25.0      EV    4193  N14902    EWR  PHL   
3426              1721      136.0      EV    4502  N15983    EWR  PHL   
10235             1722      -14.0      EV    4616  N11150    EWR  PHL   

        air_time  distance  hour  minute            time_hour  
275945       NaN        17     1       6  2013-07-27 01:00:00  
3083        30.0        80    12       0  2013-01-04 12:00:00  
3901        29.0        80    12       0  2013-01-05 12:00:00  
3426        28.0        80    16      15  2013-01-04 16:00:00  
10235       36.0        80    16      17  2013-01-12 16:00:00

## Add new columns

In [102]:
# Pandas
df = flight_data
df['new_distance'] = df['distance'] / 1000
df['carrier_origin'] = df['carrier'] + df['origin']

df

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     533.0             529        4.0     850.0   
2       2013      1    1     542.0             540        2.0     923.0   
3       2013      1    1     544.0             545       -1.0    1004.0   
4       2013      1    1     554.0             600       -6.0     812.0   
...      ...    ...  ...       ...             ...        ...       ...   
336771  2013      9   30       NaN            1455        NaN       NaN   
336772  2013      9   30       NaN            2200        NaN       NaN   
336773  2013      9   30       NaN            1210        NaN       NaN   
336774  2013      9   30       NaN            1159        NaN       NaN   
336775  2013      9   30       NaN             840        NaN       NaN   

        sched_arr_time  arr_delay carrier  ...  tailnum origin dest air_time  \
0                  819       11.0      UA  ...   N14228    EWR  IAH    227.0   
1                  830       20.0      UA  ...   N24211    LGA  IAH    227.0   
2                  850       33.0      AA  ...   N619AA    JFK  MIA    160.0   
3                 1022      -18.0      B6  ...   N804JB    JFK  BQN    183.0   
4                  837      -25.0      DL  ...   N668DN    LGA  ATL    116.0   
...                ...        ...     ...  ...      ...    ...  ...      ...   
336771            1634        0.0      9E  ...      NaN    JFK  DCA      NaN   
336772            2312        0.0      9E  ...      NaN    LGA  SYR      NaN   
336773            1330        0.0      MQ  ...   N535MQ    LGA  BNA      NaN   
336774            1344        0.0      MQ  ...   N511MQ    LGA  CLE      NaN   
336775            1020        0.0      MQ  ...   N839MQ    LGA  RDU      NaN   

        distance  hour  minute            time_hour new_distance  \
0           1400     5      15  2013-01-01 05:00:00        1.400   
1           1416     5      29  2013-01-01 05:00:00        1.416   
2           1089     5      40  2013-01-01 05:00:00        1.089   
3           1576     5      45  2013-01-01 05:00:00        1.576   
4            762     6       0  2013-01-01 06:00:00        0.762   
...          ...   ...     ...                  ...          ...   
336771       213    14      55  2013-09-30 14:00:00        0.213   
336772       198    22       0  2013-09-30 22:00:00        0.198   
336773       764    12      10  2013-09-30 12:00:00        0.764   
336774       419    11      59  2013-09-30 11:00:00        0.419   
336775       431     8      40  2013-09-30 08:00:00        0.431   

        carrier_origin  
0                UAEWR  
1                UALGA  
2                AAJFK  
3                B6JFK  
4                DLLGA  
...                ...  
336771           9EJFK  
336772           9ELGA  
336773           MQLGA  
336774           MQLGA  
336775           MQLGA  

[336776 rows x 21 columns]

## Group data

In [105]:
flight_data.groupby('origin')['distance'].agg(['mean'])

mean
origin             
EWR     1056.742790
JFK     1266.249077
LGA      779.835671

In [107]:
# Reset_index
flight_data.groupby('origin')['distance'].agg(['mean']).reset_index()

origin         mean
0    EWR  1056.742790
1    JFK  1266.249077
2    LGA   779.835671

In [109]:
# rename columns


## Group and Multiple Summarize

In [112]:
flight_data.groupby(['origin'])[['distance']].agg(['mean', 'first'])

distance      
               mean first
origin                   
EWR     1056.742790  1400
JFK     1266.249077  1089
LGA      779.835671  1416

In [114]:
# Drop index level
flight_data.groupby(['origin'])[['distance']].agg(['mean', 'first']).droplevel(axis=1,level=0)

mean  first
origin                    
EWR     1056.742790   1400
JFK     1266.249077   1089
LGA      779.835671   1416

In [116]:
# Reset_Index
flight_data.groupby(['origin'])[['distance']].agg(['mean', 'first']).droplevel(axis=1,level=0).reset_index()

origin         mean  first
0    EWR  1056.742790   1400
1    JFK  1266.249077   1089
2    LGA   779.835671   1416

## Group Two-level data

In [119]:
# Pandas
flight_data.groupby(['origin', 'dest'])['distance'].agg(['mean']).reset_index().rename(columns = {'mean' : 'mean_distance'})

origin dest  mean_distance
0      EWR  ALB          143.0
1      EWR  ANC         3370.0
2      EWR  ATL          746.0
3      EWR  AUS         1504.0
4      EWR  AVL          583.0
..     ...  ...            ...
219    LGA  SYR          198.0
220    LGA  TPA         1010.0
221    LGA  TVC          655.0
222    LGA  TYS          647.0
223    LGA  XNA         1147.0

[224 rows x 3 columns]

In [121]:
flight_data.groupby(['origin', 'dest'])['distance'].agg(['mean', 'sum']).reset_index().rename(columns = {'mean' : 'mean_distance', 'sum' : 'sum_distance'})

origin dest  mean_distance  sum_distance
0      EWR  ALB          143.0         62777
1      EWR  ANC         3370.0         26960
2      EWR  ATL          746.0       3746412
3      EWR  AUS         1504.0       1455872
4      EWR  AVL          583.0        154495
..     ...  ...            ...           ...
219    LGA  SYR          198.0         58014
220    LGA  TPA         1010.0       2166450
221    LGA  TVC          655.0         50435
222    LGA  TYS          647.0        199276
223    LGA  XNA         1147.0        854515

[224 rows x 4 columns]

## Group Two-level data & multiple summarize


In [124]:
flight_group = flight_data.groupby(['origin', 'dest'])[['distance','arr_delay']].agg(['mean', 'sum'])
flight_group

distance           arr_delay         
                mean      sum       mean      sum
origin dest                                      
EWR    ALB     143.0    62777  13.708428   6018.0
       ANC    3370.0    26960  -2.500000    -20.0
       ATL     746.0  3746412  12.848467  64525.0
       AUS    1504.0  1455872  -0.469008   -454.0
       AVL     583.0   154495   8.339623   2210.0
...              ...      ...        ...      ...
LGA    SYR     198.0    58014   8.494881   2489.0
       TPA    1010.0  2166450   8.133800  17447.0
       TVC     655.0    50435  15.142857   1166.0
       TYS     647.0   199276   3.350649   1032.0
       XNA    1147.0   854515   6.770470   5044.0

[224 rows x 4 columns]

In [126]:
# Multi indexed columns 호출
flight_group.loc[:,[('distance','mean')]]

distance
                mean
origin dest         
EWR    ALB     143.0
       ANC    3370.0
       ATL     746.0
       AUS    1504.0
       AVL     583.0
...              ...
LGA    SYR     198.0
       TPA    1010.0
       TVC     655.0
       TYS     647.0
       XNA    1147.0

[224 rows x 1 columns]

In [128]:
# Flattern Multi index
flight_group = flight_data.groupby(['origin', 'dest'])[['distance','arr_delay']].agg(['mean', 'sum']).droplevel(axis=1,level=0).reset_index()
flight_group

origin dest    mean      sum       mean      sum
0      EWR  ALB   143.0    62777  13.708428   6018.0
1      EWR  ANC  3370.0    26960  -2.500000    -20.0
2      EWR  ATL   746.0  3746412  12.848467  64525.0
3      EWR  AUS  1504.0  1455872  -0.469008   -454.0
4      EWR  AVL   583.0   154495   8.339623   2210.0
..     ...  ...     ...      ...        ...      ...
219    LGA  SYR   198.0    58014   8.494881   2489.0
220    LGA  TPA  1010.0  2166450   8.133800  17447.0
221    LGA  TVC   655.0    50435  15.142857   1166.0
222    LGA  TYS   647.0   199276   3.350649   1032.0
223    LGA  XNA  1147.0   854515   6.770470   5044.0

[224 rows x 6 columns]

In [130]:
flight_group.columns = ['origin', 'dest', 'dist_mean', 'dist_sum', 'delay_mean', 'delay_sum']
flight_group

origin dest  dist_mean  dist_sum  delay_mean  delay_sum
0      EWR  ALB      143.0     62777   13.708428     6018.0
1      EWR  ANC     3370.0     26960   -2.500000      -20.0
2      EWR  ATL      746.0   3746412   12.848467    64525.0
3      EWR  AUS     1504.0   1455872   -0.469008     -454.0
4      EWR  AVL      583.0    154495    8.339623     2210.0
..     ...  ...        ...       ...         ...        ...
219    LGA  SYR      198.0     58014    8.494881     2489.0
220    LGA  TPA     1010.0   2166450    8.133800    17447.0
221    LGA  TVC      655.0     50435   15.142857     1166.0
222    LGA  TYS      647.0    199276    3.350649     1032.0
223    LGA  XNA     1147.0    854515    6.770470     5044.0

[224 rows x 6 columns]

In [131]:
# 따로따로
a = flight_data.groupby(['origin', 'dest'])['distance'].agg(['mean']).reset_index().rename(columns = {'mean' : 'dist_mean'})
b = flight_data.groupby(['origin', 'dest'])['distance'].agg(['sum']).reset_index().rename(columns = {'sum' : 'dist_sum'})
c = flight_data.groupby(['origin', 'dest'])['arr_delay'].agg(['mean']).reset_index().rename(columns = {'mean' : 'delay_mean'})

print(a)
print(b)
print(c)

# 합치기
new_df = pd.merge(a, b)
new_df = pd.merge(new_df, c)

new_df

    origin dest  dist_mean
0      EWR  ALB      143.0
1      EWR  ANC     3370.0
2      EWR  ATL      746.0
3      EWR  AUS     1504.0
4      EWR  AVL      583.0
..     ...  ...        ...
219    LGA  SYR      198.0
220    LGA  TPA     1010.0
221    LGA  TVC      655.0
222    LGA  TYS      647.0
223    LGA  XNA     1147.0

[224 rows x 3 columns]
    origin dest  dist_sum
0      EWR  ALB     62777
1      EWR  ANC     26960
2      EWR  ATL   3746412
3      EWR  AUS   1455872
4      EWR  AVL    154495
..     ...  ...       ...
219    LGA  SYR     58014
220    LGA  TPA   2166450
221    LGA  TVC     50435
222    LGA  TYS    199276
223    LGA  XNA    854515

[224 rows x 3 columns]
    origin dest  delay_mean
0      EWR  ALB   13.708428
1      EWR  ANC   -2.500000
2      EWR  ATL   12.848467
3      EWR  AUS   -0.469008
4      EWR  AVL    8.339623
..     ...  ...         ...
219    LGA  SYR    8.494881
220    LGA  TPA    8.133800
221    LGA  TVC   15.142857
222    LGA  TYS    3.350649
223    L

origin dest  dist_mean  dist_sum  delay_mean
0      EWR  ALB      143.0     62777   13.708428
1      EWR  ANC     3370.0     26960   -2.500000
2      EWR  ATL      746.0   3746412   12.848467
3      EWR  AUS     1504.0   1455872   -0.469008
4      EWR  AVL      583.0    154495    8.339623
..     ...  ...        ...       ...         ...
219    LGA  SYR      198.0     58014    8.494881
220    LGA  TPA     1010.0   2166450    8.133800
221    LGA  TVC      655.0     50435   15.142857
222    LGA  TYS      647.0    199276    3.350649
223    LGA  XNA     1147.0    854515    6.770470

[224 rows x 5 columns]

## Mutating Categorical Data : Map and Cut


In [135]:
flight_data

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     533.0             529        4.0     850.0   
2       2013      1    1     542.0             540        2.0     923.0   
3       2013      1    1     544.0             545       -1.0    1004.0   
4       2013      1    1     554.0             600       -6.0     812.0   
...      ...    ...  ...       ...             ...        ...       ...   
336771  2013      9   30       NaN            1455        NaN       NaN   
336772  2013      9   30       NaN            2200        NaN       NaN   
336773  2013      9   30       NaN            1210        NaN       NaN   
336774  2013      9   30       NaN            1159        NaN       NaN   
336775  2013      9   30       NaN             840        NaN       NaN   

        sched_arr_time  arr_delay carrier  ...  tailnum origin dest air_time  \
0                  819       11.0      UA  ...   N14228    EWR  IAH    227.0   
1                  830       20.0      UA  ...   N24211    LGA  IAH    227.0   
2                  850       33.0      AA  ...   N619AA    JFK  MIA    160.0   
3                 1022      -18.0      B6  ...   N804JB    JFK  BQN    183.0   
4                  837      -25.0      DL  ...   N668DN    LGA  ATL    116.0   
...                ...        ...     ...  ...      ...    ...  ...      ...   
336771            1634        0.0      9E  ...      NaN    JFK  DCA      NaN   
336772            2312        0.0      9E  ...      NaN    LGA  SYR      NaN   
336773            1330        0.0      MQ  ...   N535MQ    LGA  BNA      NaN   
336774            1344        0.0      MQ  ...   N511MQ    LGA  CLE      NaN   
336775            1020        0.0      MQ  ...   N839MQ    LGA  RDU      NaN   

        distance  hour  minute            time_hour new_distance  \
0           1400     5      15  2013-01-01 05:00:00        1.400   
1           1416     5      29  2013-01-01 05:00:00        1.416   
2           1089     5      40  2013-01-01 05:00:00        1.089   
3           1576     5      45  2013-01-01 05:00:00        1.576   
4            762     6       0  2013-01-01 06:00:00        0.762   
...          ...   ...     ...                  ...          ...   
336771       213    14      55  2013-09-30 14:00:00        0.213   
336772       198    22       0  2013-09-30 22:00:00        0.198   
336773       764    12      10  2013-09-30 12:00:00        0.764   
336774       419    11      59  2013-09-30 11:00:00        0.419   
336775       431     8      40  2013-09-30 08:00:00        0.431   

        carrier_origin  
0                UAEWR  
1                UALGA  
2                AAJFK  
3                B6JFK  
4                DLLGA  
...                ...  
336771           9EJFK  
336772           9ELGA  
336773           MQLGA  
336774           MQLGA  
336775           MQLGA  

[336776 rows x 21 columns]

In [137]:
flight_data.carrier.unique()

array(['UA', 'AA', 'B6', 'DL', 'EV', 'MQ', 'US', 'WN', 'VX', 'FL', 'AS',
       '9E', 'F9', 'HA', 'YV', 'OO'], dtype=object)

In [139]:
# Mapping Carrier Class
carrier_mapping = {'UA':'A', 'AA':'A', 'B6':'A', 'DL':'A', 'EV':'A', 'MQ':'A', 'US':'A', 'WN':'B',
               'VX':'B', 'FL':'B', 'AS':'B', '9E':'B', 'F9':'B', 'HA':'B', 'YV':'B', 'OO':'B'}

flight_data['carrier_class'] = flight_data['carrier'].map(carrier_mapping)

flight_data[['carrier', 'carrier_class']].sample(n=10)

carrier carrier_class
189027      DL             A
43158       EV             A
107621      B6             A
269738      AA             A
8217        DL             A
205927      EV             A
165182      EV             A
67682       DL             A
318009      AA             A
179188      US             A

In [141]:
flight_data['carrier_class_2'] = 'A'
flight_data[['carrier', 'carrier_class','carrier_class_2']].sample(n=10)

carrier carrier_class carrier_class_2
103184      9E             B               A
193309      EV             A               A
308109      AA             A               A
247386      MQ             A               A
176738      B6             A               A
277920      AA             A               A
144786      UA             A               A
292050      UA             A               A
256081      EV             A               A
320235      B6             A               A

In [143]:
# flight_data['carrier']=='UA'이면,  'carrier_class_2' 를 'B'로 변환

flight_data.loc[flight_data['carrier']=='UA', 'carrier_class_2'] = 'B'


flight_data[['carrier', 'carrier_class', 'carrier_class_2']].sample(n=10)

carrier carrier_class carrier_class_2
82761       US             A               A
274872      EV             A               A
202093      DL             A               A
322666      EV             A               A
90892       UA             A               B
92000       EV             A               A
168919      UA             A               B
314466      MQ             A               A
269646      MQ             A               A
225295      EV             A               A

In [145]:
flight_data.distance.describe()

count    336776.000000
mean       1039.912604
std         733.233033
min          17.000000
25%         502.000000
50%         872.000000
75%        1389.000000
max        4983.000000
Name: distance, dtype: float64

In [147]:
# Caterorical variable from Continuous Variable

bins = [0,502,872,1389,5000]
labels = ["very_short", "short", "long", "very_long"]

flight_data['distance_class'] = pd.cut(flight_data['distance'], bins, labels=labels)

flight_data[['distance', 'distance_class']].sample(n=10)

distance distance_class
241232      2446      very_long
15892        529          short
101667      2586      very_long
172334       963           long
69747        529          short
276867       544          short
122917      1092           long
48568       2586      very_long
86136         96     very_short
212919       950           long

## 연습문제 02

* \[Step 1\]: Filter out all flights less than 10 hours
* \[Step 2\]: Create a new column, `speed`, using the formula [distance / (air time * 60)]
* \[Step 3\]: Calculate the mean speed for flights originating from each airport
* \[Step 4\]: Sort the result by mean speed in descending order

In [150]:
result = flight_data.loc[flight_data['hour'] > 10, :]
result['speed'] = result['distance'] / (result['air_time'] * 60)
result = result.groupby('origin')['speed'].agg(['mean']).reset_index().rename(columns = {'mean' : 'mean_speed'})
result = result.sort_values('mean_speed', ascending=False)

result

C:\Users\leeja\AppData\Local\Temp\ipykernel_19592\1241633039.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  result['speed'] = result['distance'] / (result['air_time'] * 60)


origin  mean_speed
0    EWR    0.109777
1    JFK    0.109427
2    LGA    0.107362

## From Wide to Long

In [153]:
np.random.seed(123)
df = pd.DataFrame({"A1970" : {0 : "a", 1 : "b", 2 : "c"},
                   "A1980" : {0 : "d", 1 : "e", 2 : "f"},
                   "B1970" : {0 : 2.5, 1 : 1.2, 2 : .7},
                   "B1980" : {0 : 3.2, 1 : 1.3, 2 : .1},
                   "X"     : dict(zip(range(3), np.random.randn(3)))
                  })
df["id"] = df.index
df

A1970 A1980  B1970  B1980         X  id
0     a     d    2.5    3.2 -1.085631   0
1     b     e    1.2    1.3  0.997345   1
2     c     f    0.7    0.1  0.282978   2

In [155]:
pd.wide_to_long(df, ["A", "B"], i="id", j="year")

X  A    B
id year                  
0  1970 -1.085631  a  2.5
1  1970  0.997345  b  1.2
2  1970  0.282978  c  0.7
0  1980 -1.085631  d  3.2
1  1980  0.997345  e  1.3
2  1980  0.282978  f  0.1

In [157]:
df = pd.DataFrame({
    'famid': [1, 1, 1, 2, 2, 2, 3, 3, 3],
    'birth': [1, 2, 3, 1, 2, 3, 1, 2, 3],
    'ht1': [2.8, 2.9, 2.2, 2, 1.8, 1.9, 2.2, 2.3, 2.1],
    'ht2': [3.4, 3.8, 2.9, 3.2, 2.8, 2.4, 3.3, 3.4, 2.9]
})
df

famid  birth  ht1  ht2
0      1      1  2.8  3.4
1      1      2  2.9  3.8
2      1      3  2.2  2.9
3      2      1  2.0  3.2
4      2      2  1.8  2.8
5      2      3  1.9  2.4
6      3      1  2.2  3.3
7      3      2  2.3  3.4
8      3      3  2.1  2.9

In [159]:
# With multiple id columns

pd.wide_to_long(df, stubnames='ht', i=['famid', 'birth'], j='age').reset_index()

famid  birth  age   ht
0       1      1    1  2.8
1       1      1    2  3.4
2       1      2    1  2.9
3       1      2    2  3.8
4       1      3    1  2.2
5       1      3    2  2.9
6       2      1    1  2.0
7       2      1    2  3.2
8       2      2    1  1.8
9       2      2    2  2.8
10      2      3    1  1.9
11      2      3    2  2.4
12      3      1    1  2.2
13      3      1    2  3.3
14      3      2    1  2.3
15      3      2    2  3.4
16      3      3    1  2.1
17      3      3    2  2.9

In [161]:
df = pd.DataFrame({
    'date' : ['05/03', '06/03', '07/03', '08/03'],
    'AA' : [1, 4, 7, 5],
    'BB' : [2, 5, 8, 7],
    'CC' : [3, 6, 9, 1]
}).set_index('date')

df = df.reset_index()
df

date  AA  BB  CC
0  05/03   1   2   3
1  06/03   4   5   6
2  07/03   7   8   9
3  08/03   5   7   1

In [163]:
# Use pd.melt

pd.melt(df, id_vars='date', value_vars=['AA', 'BB', 'CC'])

date variable  value
0   05/03       AA      1
1   06/03       AA      4
2   07/03       AA      7
3   08/03       AA      5
4   05/03       BB      2
5   06/03       BB      5
6   07/03       BB      8
7   08/03       BB      7
8   05/03       CC      3
9   06/03       CC      6
10  07/03       CC      9
11  08/03       CC      1

## Join

In [180]:
# Colab 사용
#flight_data = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/5_스마트도시데이터분석/codes/lecture02_data_processing/data/nycflights13.csv')

# Jupyter Notebook 사용
flight_data = pd.read_csv('data/nycflights13.csv')

flight_data_reset  = flight_data

flight_data_reset

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     533.0             529        4.0     850.0   
2       2013      1    1     542.0             540        2.0     923.0   
3       2013      1    1     544.0             545       -1.0    1004.0   
4       2013      1    1     554.0             600       -6.0     812.0   
...      ...    ...  ...       ...             ...        ...       ...   
336771  2013      9   30       NaN            1455        NaN       NaN   
336772  2013      9   30       NaN            2200        NaN       NaN   
336773  2013      9   30       NaN            1210        NaN       NaN   
336774  2013      9   30       NaN            1159        NaN       NaN   
336775  2013      9   30       NaN             840        NaN       NaN   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
0                  819       11.0      UA    1545  N14228    EWR  IAH   
1                  830       20.0      UA    1714  N24211    LGA  IAH   
2                  850       33.0      AA    1141  N619AA    JFK  MIA   
3                 1022      -18.0      B6     725  N804JB    JFK  BQN   
4                  837      -25.0      DL     461  N668DN    LGA  ATL   
...                ...        ...     ...     ...     ...    ...  ...   
336771            1634        NaN      9E    3393     NaN    JFK  DCA   
336772            2312        NaN      9E    3525     NaN    LGA  SYR   
336773            1330        NaN      MQ    3461  N535MQ    LGA  BNA   
336774            1344        NaN      MQ    3572  N511MQ    LGA  CLE   
336775            1020        NaN      MQ    3531  N839MQ    LGA  RDU   

        air_time  distance  hour  minute            time_hour  
0          227.0      1400     5      15  2013-01-01 05:00:00  
1          227.0      1416     5      29  2013-01-01 05:00:00  
2          160.0      1089     5      40  2013-01-01 05:00:00  
3          183.0      1576     5      45  2013-01-01 05:00:00  
4          116.0       762     6       0  2013-01-01 06:00:00  
...          ...       ...   ...     ...                  ...  
336771       NaN       213    14      55  2013-09-30 14:00:00  
336772       NaN       198    22       0  2013-09-30 22:00:00  
336773       NaN       764    12      10  2013-09-30 12:00:00  
336774       NaN       419    11      59  2013-09-30 11:00:00  
336775       NaN       431     8      40  2013-09-30 08:00:00  

[336776 rows x 19 columns]

In [182]:

# index를 활용해서 id 컬럼 생성

flight_data_reset.reset_index(inplace=True)
flight_data_reset=flight_data_reset.rename(columns = {'index' : 'id'})
flight_data_reset['id']=flight_data_reset['id']+1
flight_data_reset['id']=flight_data_reset['id'].astype('int')

print(flight_data_reset.dtypes)
flight_data_reset

id                  int32
year                int64
month               int64
day                 int64
dep_time          float64
sched_dep_time      int64
dep_delay         float64
arr_time          float64
sched_arr_time      int64
arr_delay         float64
carrier            object
flight              int64
tailnum            object
origin             object
dest               object
air_time          float64
distance            int64
hour                int64
minute              int64
time_hour          object
dtype: object


id  year  month  day  dep_time  sched_dep_time  dep_delay  \
0            1  2013      1    1     517.0             515        2.0   
1            2  2013      1    1     533.0             529        4.0   
2            3  2013      1    1     542.0             540        2.0   
3            4  2013      1    1     544.0             545       -1.0   
4            5  2013      1    1     554.0             600       -6.0   
...        ...   ...    ...  ...       ...             ...        ...   
336771  336772  2013      9   30       NaN            1455        NaN   
336772  336773  2013      9   30       NaN            2200        NaN   
336773  336774  2013      9   30       NaN            1210        NaN   
336774  336775  2013      9   30       NaN            1159        NaN   
336775  336776  2013      9   30       NaN             840        NaN   

        arr_time  sched_arr_time  arr_delay carrier  flight tailnum origin  \
0          830.0             819       11.0      UA    1545  N14228    EWR   
1          850.0             830       20.0      UA    1714  N24211    LGA   
2          923.0             850       33.0      AA    1141  N619AA    JFK   
3         1004.0            1022      -18.0      B6     725  N804JB    JFK   
4          812.0             837      -25.0      DL     461  N668DN    LGA   
...          ...             ...        ...     ...     ...     ...    ...   
336771       NaN            1634        NaN      9E    3393     NaN    JFK   
336772       NaN            2312        NaN      9E    3525     NaN    LGA   
336773       NaN            1330        NaN      MQ    3461  N535MQ    LGA   
336774       NaN            1344        NaN      MQ    3572  N511MQ    LGA   
336775       NaN            1020        NaN      MQ    3531  N839MQ    LGA   

       dest  air_time  distance  hour  minute            time_hour  
0       IAH     227.0      1400     5      15  2013-01-01 05:00:00  
1       IAH     227.0      1416     5      29  2013-01-01 05:00:00  
2       MIA     160.0      1089     5      40  2013-01-01 05:00:00  
3       BQN     183.0      1576     5      45  2013-01-01 05:00:00  
4       ATL     116.0       762     6       0  2013-01-01 06:00:00  
...     ...       ...       ...   ...     ...                  ...  
336771  DCA       NaN       213    14      55  2013-09-30 14:00:00  
336772  SYR       NaN       198    22       0  2013-09-30 22:00:00  
336773  BNA       NaN       764    12      10  2013-09-30 12:00:00  
336774  CLE       NaN       419    11      59  2013-09-30 11:00:00  
336775  RDU       NaN       431     8      40  2013-09-30 08:00:00  

[336776 rows x 20 columns]

In [184]:
# id 컬럼을 공유하는 두 개의 데이터 프레임 생성
flight_distance = flight_data_reset.loc[flight_data_reset['carrier'].isin(['DL','AA']), ['id','carrier','distance']] # DL, AA
flight_time = flight_data_reset.loc[flight_data_reset['carrier'].isin(['AA']), ['id','carrier','air_time']] #  AA 만

In [186]:
flight_distance

id carrier  distance
2            3      AA      1089
4            5      DL       762
9           10      AA       733
14          15      AA      1389
20          21      DL      1020
...        ...     ...       ...
336709  336710      AA       733
336715  336716      DL       187
336718  336719      DL      1598
336744  336745      DL      2475
336751  336752      AA      2475

[80839 rows x 3 columns]

In [188]:
flight_time

id carrier  air_time
2            3      AA     160.0
9           10      AA     138.0
14          15      AA     257.0
22          23      AA     152.0
31          32      AA     153.0
...        ...     ...       ...
336653  336654      AA     150.0
336700  336701      AA      37.0
336704  336705      AA     294.0
336709  336710      AA     105.0
336751  336752      AA     298.0

[32729 rows x 3 columns]

병합 방식은 크게 inner, outer, left, right 방식이 있습니다.
* inner join 방식 : 좌측 table과 우측 table의 기준열에서 key 값이 모두 존재하는 경우에만 병합
* outer join 방식 : 좌측 table과 우측 table의 기준열에서 key 값이 모두 존재하지 않아도 병합
* left join 방식 : 좌측 table의 기준열만 고려하여 병합
* right join 방식 : 우측 table의 기준열만 고려하여 병합

여기서는 많이 사용하는 inner join, left join에 대해서만 배우겠습니다.

<img src="https://static.wixstatic.com/media/2f98e9_c5999377496a4e399bbc042ffe3280fc~mv2.jpg" width="600" height="400" />

In [192]:
# 왼쪽의 기준 Table을 먼저 쓰고, 오른쪽의 병합 Table을 다음에 씁니다. left_on과 right_on은 각각 Table의 기준열입니다.
#inner join 방식입니다. inner join이  default 설정이기 때문에 how = 'inner' parameter를 넣지 않아도 됩니다.

flight_inner = pd.merge(flight_distance, flight_time, left_on='id', right_on="id", how='inner')

flight_inner

id carrier_x  distance carrier_y  air_time
0           3        AA      1089        AA     160.0
1          10        AA       733        AA     138.0
2          15        AA      1389        AA     257.0
3          23        AA      1085        AA     152.0
4          32        AA      1096        AA     153.0
...       ...       ...       ...       ...       ...
32724  336654        AA      1096        AA     150.0
32725  336701        AA       187        AA      37.0
32726  336705        AA      2475        AA     294.0
32727  336710        AA       733        AA     105.0
32728  336752        AA      2475        AA     298.0

[32729 rows x 5 columns]

flight_distance은 80839개의 행을 가지고 있었으나, inner join은 'flight_distance' 'flight_time'에 공통적으로 포함된 값들만 살리고, 이들을 기준으로 데이터를 병합합니다. 따라서 flight_distance에서 carrier가 AA가 아닌 행들은 모두 삭제됩니다.

In [195]:
#left join 방식입니다.
flight_left = pd.merge(flight_distance, flight_time, left_on='id', right_on="id", how='left')

flight_left

id carrier_x  distance carrier_y  air_time
0           3        AA      1089        AA     160.0
1           5        DL       762       NaN       NaN
2          10        AA       733        AA     138.0
3          15        AA      1389        AA     257.0
4          21        DL      1020       NaN       NaN
...       ...       ...       ...       ...       ...
80834  336710        AA       733        AA     105.0
80835  336716        DL       187       NaN       NaN
80836  336719        DL      1598       NaN       NaN
80837  336745        DL      2475       NaN       NaN
80838  336752        AA      2475        AA     298.0

[80839 rows x 5 columns]

left join은 왼쪽의 기준테이블의 모든행을 유지합니다. 따라서 flight_distance은 80839개의 행을 모두 유지합니다. 기준테이블과 결합테이블 간에 일치하지 않는 행은 결측치가 입력됩니다.

## Stacking

In [199]:
# @title 기본 제목 텍스트
df1 = pd.DataFrame([['A0', 'A1', 'A2', 'A3'],
                    ['B0', 'B1', 'B2', 'B3'],
                    ['C0', 'C1', 'C2', 'C3'],
                    ['D0', 'D1', 'D2', 'D3']], columns=list('ABCD'))

df2 = pd.DataFrame([['A4', 'A5', 'A6', 'A7'],
                    ['B4', 'B5', 'B6', 'B7'],
                    ['C4', 'C5', 'C6', 'C7'],
                    ['D4', 'D5', 'D6', 'D7']], columns=['A', 'B', 'C', 'D'], index=[4, 5, 6, 7])

df3 = pd.DataFrame([['A8', 'A9', 'A10', 'A11'],
                    ['B8', 'B9', 'B10', 'B11'],
                    ['C8', 'C9', 'C10', 'C11'],
                    ['D8', 'D9', 'D10', 'D11']], columns=list('ABCD'), index=[8, 9, 10, 11])

In [201]:
df1

A   B   C   D
0  A0  A1  A2  A3
1  B0  B1  B2  B3
2  C0  C1  C2  C3
3  D0  D1  D2  D3

In [203]:
df2

A   B   C   D
4  A4  A5  A6  A7
5  B4  B5  B6  B7
6  C4  C5  C6  C7
7  D4  D5  D6  D7

In [205]:
df3

A   B    C    D
8   A8  A9  A10  A11
9   B8  B9  B10  B11
10  C8  C9  C10  C11
11  D8  D9  D10  D11

In [207]:
# 열방향 병합
pd.concat([df1, df2, df3])

A   B    C    D
0   A0  A1   A2   A3
1   B0  B1   B2   B3
2   C0  C1   C2   C3
3   D0  D1   D2   D3
4   A4  A5   A6   A7
5   B4  B5   B6   B7
6   C4  C5   C6   C7
7   D4  D5   D6   D7
8   A8  A9  A10  A11
9   B8  B9  B10  B11
10  C8  C9  C10  C11
11  D8  D9  D10  D11

In [209]:
# axis=1 옵션을 통한 행 방향 병합
pd.concat([df1, df2, df3], axis=1)

A    B    C    D    A    B    C    D    A    B    C    D
0    A0   A1   A2   A3  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1    B0   B1   B2   B3  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2    C0   C1   C2   C3  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
3    D0   D1   D2   D3  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
4   NaN  NaN  NaN  NaN   A4   A5   A6   A7  NaN  NaN  NaN  NaN
5   NaN  NaN  NaN  NaN   B4   B5   B6   B7  NaN  NaN  NaN  NaN
6   NaN  NaN  NaN  NaN   C4   C5   C6   C7  NaN  NaN  NaN  NaN
7   NaN  NaN  NaN  NaN   D4   D5   D6   D7  NaN  NaN  NaN  NaN
8   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   A8   A9  A10  A11
9   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   B8   B9  B10  B11
10  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   C8   C9  C10  C11
11  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   D8   D9  D10  D11